packages/medusa/src/api/routes/admin/products/create-variant.ts

```diff
@@ -111,18 +169,139 @@ export default async (req, res) => {
     req.body
   )
 
+  const inventoryService: IInventoryService =
+    req.scope.resolve("inventoryService")
+  const productVariantInventoryService: ProductVariantInventoryService =
+    req.scope.resolve("productVariantInventoryService")
   const productVariantService: ProductVariantService = req.scope.resolve(
     "productVariantService"
   )
-  const productService: ProductService = req.scope.resolve("productService")
+
+  const createdId: Record<string, string | null> = {
+    variant: null,
+    inventoryItem: null,
+  }
 
   const manager: EntityManager = req.scope.resolve("manager")
   await manager.transaction(async (transactionManager) => {
-    return await productVariantService
-      .withTransaction(transactionManager)
-      .create(id, validated as CreateProductVariantInput)
+    const inventoryServiceTx =
+      inventoryService?.withTransaction?.(transactionManager)
+    const productVariantInventoryServiceTx =
+      productVariantInventoryService?.withTransaction?.(transactionManager)
+
+    const productVariantServiceTx =
+      productVariantService.withTransaction(transactionManager)
+
+    async function createVariant() {
+      const variant = await productVariantServiceTx.create(
+        id,
+        validated as CreateProductVariantInput
+      )
+
+      createdId.variant = variant.id
+
+      return { variant }
+    }
+
+    async function removeVariant() {
+      if (createdId.variant) {
+        await productVariantServiceTx.delete(createdId.variant)
+      }
```

Comment:
Question: can this receive the successful result of the same step for the compensation action? 

I was thinking this might be a good pattern to introduce, or at least make possible in order to not rely on local state outside the transaction orchestrator. I dont think it should be necessary to include in this pr 😄

Reply:
It's planned to have a feature other than `forwardResponse` that will store all the responses, so steps will have access to all the responses of all its ancestors. That will help the `compensation` access the response of the `invoke` from the same action.